dali/operators/util/randomizer.cu

```diff
@@ -14,29 +14,35 @@
 
 #include "dali/operators/util/randomizer.cuh"
 
-
 namespace dali {
 
+namespace detail {
+
 __global__
-void initializeStates(const int N, unsigned int seed, curandState *states) {
-  for (int idx = threadIdx.x + blockIdx.x * blockDim.x;
+void init_states(const size_t N, uint64_t seed, curandState *states) {
+  for (size_t idx = threadIdx.x + blockIdx.x * blockDim.x;
        idx < N;
        idx += blockDim.x * gridDim.x) {
     curand_init(seed, idx, 0, &states[idx]);
   }
 }
 
-RandomizerGPU::RandomizerGPU(int seed, size_t len) {
-  len_ = len;
+}  // namespace detail
+
+curand_states::curand_states(uint64_t seed, size_t len) : len_(len) {
   cudaGetDevice(&device_);
-  states_ = static_cast<curandState*>(GPUBackend::New(sizeof(curandState) * len, true));
-  initializeStates<<<div_ceil(len, block_size_), block_size_>>>
-                  (len_, seed, reinterpret_cast<curandState*>(states_));
+  states_mem_ = kernels::memory::alloc_unique<curandState>(kernels::AllocType::GPU, len);
+  states_ = states_mem_.get();
+  static constexpr int kBlockSize = 256;
+  int grid = div_ceil(len_, kBlockSize);
+  detail::init_states<<<grid, kBlockSize>>>(len_, seed, states_);
 }
 
-void RandomizerGPU::Cleanup() {
+curand_states::~curand_states() {
+  // TODO(janton) do we need this?
```

Comment:
I don't think it is. From my experience, cudaFree can be called on any device pointer.

Reply:
Update: 64-bit applications are running under Unified Virtual Addressing and pointers are unique system-wide. `cudaFree` correctly identifies the owning device. I'm triple-checking if it covers AARCH64 and POWER, but I see no reason not to.